src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -2687,6 +2687,20 @@ const Type* CatchNode::Value(PhaseGVN* phase) const {
       // Rethrows always throw exceptions, never return
       if (call->entry_point() == OptoRuntime::rethrow_stub()) {
         f[CatchProjNode::fall_through_index] = Type::TOP;
+      } else if (call->is_AllocateArray()) {
+        Node* klass_node = call->in(AllocateNode::KlassNode);
+        Node *length = call->in(AllocateNode::ALength);
+        const Type* length_type = phase->type(length);
+        const Type* klass_type = phase->type(klass_node);
+        if (length_type == Type::TOP || klass_type == Type::TOP) {
+          f[CatchProjNode::fall_through_index] = Type::TOP;
+        } else {
+          Node* valid_length_test = call->in(AllocateNode::ValidLengthTest);
+          const Type* valid_length_test_t = phase->type(valid_length_test);
+          if (valid_length_test_t->isa_int() && valid_length_test_t->is_int()->is_con(0)) {
```

Comment:
Here you do: 
```
Node* valid_length_test = call->in(AllocateNode::ValidLengthTest);
const Type* valid_length_test_t = phase->type(valid_length_test);
if (valid_length_test_t->isa_int() && valid_length_test_t->is_int()->is_con(0)) {
```

But in compile.cpp:3766 you do:
```
Node* valid_length_test = call->in(call->req());
call->rm_prec(call->req());
if (valid_length_test->find_int_con(1) == 0) {
```

Why "call->req()" and not "call->in(AllocateNode::ValidLengthTest)"?

And why not "if (valid_length_test->find_int_con(1) == 0) {" in both places?

Reply:
> Here you do:
> 
> ```
> Node* valid_length_test = call->in(AllocateNode::ValidLengthTest);
> const Type* valid_length_test_t = phase->type(valid_length_test);
> if (valid_length_test_t->isa_int() && valid_length_test_t->is_int()->is_con(0)) {
> ```
> 
> But in compile.cpp:3766 you do:
> 
> ```
> Node* valid_length_test = call->in(call->req());
> call->rm_prec(call->req());
> if (valid_length_test->find_int_con(1) == 0) {
> ```
> 
> Why "call->req()" and not "call->in(AllocateNode::ValidLengthTest)"?

call->in(AllocateNode::ValidLengthTest) only works for allocateArrayNode. The code I modified in compile.cpp runs after macro expansion so there's no AllocateArrayNode anymore. Instead there's a call to the runtime. When the AllocateArrayNode is macro expanded I move in(AllocateNode::ValidLengthTest) to the new runtime call as a precedence edge that is at req(). I tried adding it as an extra parameter that would be removed in compile.cpp but that messed up debug infos and I hit some asserts.

> > And why not "if (valid_length_test->find_int_con(1) == 0) {" in both places?

So`that the Value() call does the right thing during CCP where a node can be registered as a constant but not transformed yet to a ConNode.

Reply:
That's make sense. Thanks for explaining!